### DISCUSSION TITLE
sfnetworks  for aggregating routes to flows

### DISCUSSION THREAD

**Discussion by shriv** [luukvdmeer/sfnetworks#149] [Show and tell]
Hi  everyone, 
I'm a reasonably new user of this excellent package. My particular use case (for the moment) is creating visualisations of flows on a transport network. I've got a simple example of port to port rail flows written up [here](https://shriv-portfolio.netlify.app/post/infrastructure-flows/). Appreciate any thoughts, comments and critique :-)

![image](https://user-images.githubusercontent.com/3827179/115690127-d9e27280-a3b0-11eb-8ae4-f698ca9c4a4e.png)

Cheers! 
Shrividya

**Comment by Robinlovelace** [comment]
Hi @shriv this is a good question and I've been wondering the same thing. Related issue: https://github.com/ropensci/stplanr/issues/458

Documentation of another way of generating flows, using the `overline()` function in `stplanr`: https://docs.ropensci.org/stplanr/reference/overline.html

Hypothesis: it can be done faster and more flexibly with `sfnetworks`. Will be great if so and I've been wanting to get it working for some time now as discussed months back with @agila5. One question: do you have a reproducible example? You could start with a simple dataset like this, can you reproduce the results and is this what you mean:

```
remotes::install_cran("stplanr")
library(stplanr)
sl = routes_fast_sf[2:4, ]
sl$All = flowlines$All[2:4]
rnet = overline(sl = sl, attrib = "All")
nrow(sl)
nrow(rnet)
plot(rnet)
```

**Reply by shriv** [reply]
Hi @Robinlovelace, 
I have a "solution" for aggregating routes to flows with `sfnetworks` written up [here](https://github.com/shriv/r-geospatial/blob/master/R/infrastructure-flows.Rmd). It's basically the executable file behind the blog post I linked in my first comment. :-)
I haven't done any speed tests or anything at the moment and I am keen to learn if this approach I've taken is sensible!

**Reply by Robinlovelace** [reply]
Great to see some reproducible code. Could you try comparing the output in your blogpost with output in generated by `overline()`? I would like to use `sfnetworks` as a back-end in `stplanr` for local routing and this could be a good motivation for moving that forward.

**Reply by shriv** [reply]
I can definitely give it a go! :-)

**Reply by agila5** [reply]
Hi @shriv! First of all, thank you very much for the blog posts (I loved reading them) and for the suggestions on how to aggregate OD flows using `sfnetworks`. I will also check your proposal as soon as possible. For the moment I have just one suggestion: you can use the [`to_spatial_shortest_path`](https://luukvdmeer.github.io/sfnetworks/articles/morphers.html#to-spatial-shortest-paths) morpher to extract the segments that belong to the shortest path and skip the `st_nearest_feature` step. 

> Hypothesis: it can be done faster and more flexibly with sfnetworks.

I'm not 100% sure but I think that, at the moment, there is no "support" for overlapping line segments in `sfnetworks`, so I think that `overline` answers a different question.

**Reply by shriv** [reply]
Hi @agila5, 
Thanks for the feedback! Thinking about the `to_spatial_shortest_path`, not sure it will help? I have a separate set of origin-destination points which I'm trying to route between. And afaik, I need to find the closest nodes to them first (before routing). I think the `to_spatial_shortest_path` will help for some other analyses I have in mind - where I try to find "arterial" routes to key locations. :-)

**Comment by loreabad6** [comment]
Hi @shriv! First of all thank you for pointing us to your blogpost! It is a super nice example of `sfnetworks` in the wild, I really enjoyed it. 

From a strictly dataviz perspective (so not commenting on code optimization here 😅) I have a couple of points:

- About assigning flows to the size aesthetic:
    - I find it particularly hard to differentiate between the edges with flow 100 and those with 500 and in between. The range could be wider, starting from a narrower linewidth to something thicker.
    - Since a spatial network has edges "embedded" in space, the line drawing in ggplot2 gets complicated considering curves, if you look closely there are partitions between the lines specially when the width of the line is thicker, which is harsh on the eye when trying to give a message of a continuous rail section. So in this point I would say probably using a not too large range is optimal.
    -  Finally, since sticking to a short range will lead us to the same problem of not differentiating between flow breaks, I suggest to combine it with a color legend. That way you still give more importance to the higher flows, but also show the differences clearly with color.

- About the port point locations: 
    -  Since most of the "original" and "nearest" port locations are very close to each other due to the map scale, it comes to the point where some ports are simply not visible. 
    -  For this I would suggest playing with the order of the layers and the sizes.
    - I assume that the ports you plot in grey are less important than those in red (and this assumption is merely based on the color choice), so I would plot those as base, and on top the red ones with a smaller size so you can actually see them.

On top of this. I know that it is tricky to use ggplot with several color scales, i.e. giving a color legend for the points and another for the lines is not possible, unless some workarounds are used. Some options:
1. use `ggnewscale` to assign two different geoms to two different color legends. 
2. use a different shape for the points (specifically `shape = 21`), where you assign the points to the fill aesthetic.
3. test your options with `ggraph`. If you actually end up with one single `sfnetwork` that contains as much information about flows and ports, then you can use a combination of ggraph and geom_sf to achieve your goals. The integration of both packages is not perfect yet, but [I wrote a blogpost](https://loreabad6.github.io/posts/2021-03-05-geospatial-network-visualization/) with some hints that might be useful for this or your future work.

I played a bit with your code to generate the plot you attached to integrate the recommendations I listed here. 
Code: 
```r
ggplot() + 
  geom_sf(data = railway_net %>%
            activate("edges") %>% 
            st_as_sf(),
          size = 0.1) + 
  geom_sf(data = railway_net %>% 
            activate("edges") %>%
            st_as_sf() %>% 
            mutate(row = row_number()) %>% 
            inner_join(flows_df, by = c("row" = "edge_paths")),
          aes(colour = flow, size = flow)) + 
  scale_colour_viridis_c(
    "Flow", limits = c(0,550), breaks = seq(100, 600, 100)
  ) +
  scale_size_continuous(
    "Flow", range = c(0.2, 1.25), breaks = seq(100, 600, 100),
    limits = c(0,550)
  ) +
  guides(colour = guide_legend(), size = guide_legend()) +
  geom_sf(data = orig_dest %>% select(from_port) %>% distinct() %>% st_as_sf(),
          aes(fill = "grey60"),
          size = 4, stroke = 0.01, shape = 21) +
  geom_sf(data = railway_net %>%
            activate("nodes") %>%
            st_as_sf() %>%
            mutate(row = row_number()) %>%
            inner_join(orig_dest, by = c("row" = "to_index")),
          aes(fill = "red"),
          size = 2.5, stroke = 0.01, shape = 21) +
  scale_fill_identity(breaks = c("red", "grey60"),
                        labels = c("nearest rail port", "original port"),
                        guide = "legend",
                        name = "")
```
Plot:
![flow_example](https://user-images.githubusercontent.com/10034237/116820749-e1ef9e80-ab76-11eb-8ac3-4bbb1c93bbbe.png)

I hope this is useful for you! I am very happy to keep exchanging ideas on your work and to hear your opinions too 😄

**Reply by shriv** [reply]
Hi @loreabad6 thank you for this! I really appreciate your comments. Apologies for the long delay in response - had a bunch of new project kick off craziness that has only calmed down now! :-) I like the colour suggestions! I do also like the thickness visualisation as it has the advantage of like-for-like physical to digital representation. For example, when people talk about 'heavy traffic flows', that language symbology maps well to heavy line thickness. At least for me it does! I agree with the inability to extract useful information from the thickness scales though. I wonder if some kind of parametric scaling might help? Maybe a logistic scale factor or some other parametric function to convert a flow value to a scaled one that is more informative in the visualisation? Do you have any suggestions if this approach would work / be sensible even?

Thank you!!
